lib/join/server_env0.go

```diff
@@ -0,0 +1,129 @@
+// Teleport
+// Copyright (C) 2025 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package join
+
+import (
+	"context"
+
+	"github.com/gravitational/trace"
+
+	workloadidentityv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/workloadidentity/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/join/env0"
+	"github.com/gravitational/teleport/lib/join/internal/authz"
+	"github.com/gravitational/teleport/lib/join/internal/diagnostic"
+	"github.com/gravitational/teleport/lib/join/internal/messages"
+)
+
+type Env0TokenValidator interface {
+	ValidateToken(
+		ctx context.Context,
+		token []byte,
+	) (*env0.IDTokenClaims, error)
+}
+
+// handleTokenJoin handles join attempts for the token join method.
+func (s *Server) handleEnv0Join(
```

Comment:
nit: the comment has copy/paste error

but, I think everything in this method except the token validator and the allow rule checker will be common for all OIDC join methods

what would you think of having a single handleOIDCJoin method that accepts the token validator and allow rule checker as an argument, maybe this signature

```go
// oidcTokenValidator is a function type that validates an OIDC token and checks that
// it matches an allow rule configured in the provision token.
type oidcTokenValidator func(
	ctx context.Context,
	provisionToken types.ProvisionToken,
	idToken []byte,
) (rawClaims any, joinAttrs *workloadidentityv1.JoinAttrs, err error)
```

to reduce duplication and separate the concerns of handling the message stream from actually checking the token